### PR TITLE
Fix the issue for AquilaChat2-* models

### PR DESCRIPTION
**Pull request by lu-wang-dl**
The architectures  of `BAAI/AquilaChat2-34B` and `BAAI/AquilaChat2-7B` are `AquilaForCausalLM`. When loading the model, it will show error:
```
ValueError: Model architectures ['AquilaForCausalLM'] are not supported for now. Supported architectures: ['AquilaModel', 'BaiChuanForCausalLM', 'BaichuanForCausalLM', 'BloomForCausalLM', 'FalconForCausalLM', 'GPT2LMHeadModel', 'GPTBigCodeForCausalLM', 'GPTJForCausalLM', 'GPTNeoXForCausalLM', 'InternLMForCausalLM', 'LlamaForCausalLM', 'LLaMAForCausalLM', 'MistralForCausalLM', 'MPTForCausalLM', 'OPTForCausalLM', 'QWenLMHeadModel', 'RWForCausalLM']
```

This PR is to add the arch name and fix the issue.

Manual test with the new and old models:
Screenshot 2023-10-13 at 12 02 13 AM
Screenshot 2023-10-13 at 1 17 34 AM
Screenshot 2023-10-13 at 1 43 42 AM

### CODE REVIEW

**Comment by gesanqiu** [comment]
@lu-wang-dl Have you been test your model? Even the tokenizer file updated on the HF is broken, can't load correctly.
```
from transformers import AutoTokenizer, AutoModelForCausalLM, AutoModel, CodeLlamaTokenizer
import torch

model_path = "BAAI/AquilaChat2-34B"
tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
model = AutoModelForCausalLM.from_pretrained(model_path, trust_remote_code=True)
````
```
(vllm) dell@dell:~/workSpace$ python test/download.py
Traceback (most recent call last):
  File "test/download.py", line 10, in <module>
    tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
  File "/home/dell/anaconda3/envs/vllm/lib/python3.8/site-packages/transformers/models/auto/tokenization_auto.py", line 751, in from_pretrained
    return tokenizer_class.from_pretrained(pretrained_model_name_or_path, *inputs, **kwargs)
  File "/home/dell/anaconda3/envs/vllm/lib/python3.8/site-packages/transformers/tokenization_utils_base.py", line 2045, in from_pretrained
    return cls._from_pretrained(
  File "/home/dell/anaconda3/envs/vllm/lib/python3.8/site-packages/transformers/tokenization_utils_base.py", line 2256, in _from_pretrained
    tokenizer = cls(*init_inputs, **init_kwargs)
  File "/home/dell/anaconda3/envs/vllm/lib/python3.8/site-packages/transformers/models/gpt2/tokenization_gpt2_fast.py", line 136, in __init__
    super().__init__(
  File "/home/dell/anaconda3/envs/vllm/lib/python3.8/site-packages/transformers/tokenization_utils_fast.py", line 111, in __init__
    fast_tokenizer = TokenizerFast.from_file(fast_tokenizer_file)
Exception: EOF while parsing a value at line 156568 column 3
```

**Comment by gesanqiu** [comment]
> Only add this line seems not enough, It will get a new `Assertion Error` at `assert param_slice.shape == loaded_weight.shape` (line 370 from `aquila.py`)

I believe this problem is because the shape of the tensor is wrong after tp? You can test this model on A100 80G, or add some log in `load_padded_tensor_parallel_vocab` or `load_tensor_parallel_weights` where the error happened, and modify the related parameters in config.json to get a quick fix.

**Comment by ftgreat** [comment]
> > @lu-wang-dl Have you been test your model? Even the tokenizer file you updated on the HF is broken, can't load correctly.
> > ```
> > from transformers import AutoTokenizer, AutoModelForCausalLM, AutoModel, CodeLlamaTokenizer
> > import torch
> > 
> > model_path = "BAAI/AquilaChat2-34B"
> > tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
> > model = AutoModelForCausalLM.from_pretrained(model_path, trust_remote_code=True)
> > ```
> > 
> > 
> > ```
> > (vllm) dell@dell:~/workSpace$ python test/download.py
> > Traceback (most recent call last):
> >   File "test/download.py", line 10, in <module>
> >     tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
> >   File "/home/dell/anaconda3/envs/vllm/lib/python3.8/site-packages/transformers/models/auto/tokenization_auto.py", line 751, in from_pretrained
> >     return tokenizer_class.from_pretrained(pretrained_model_name_or_path, *inputs, **kwargs)
> >   File "/home/dell/anaconda3/envs/vllm/lib/python3.8/site-packages/transformers/tokenization_utils_base.py", line 2045, in from_pretrained
> >     return cls._from_pretrained(
> >   File "/home/dell/anaconda3/envs/vllm/lib/python3.8/site-packages/transformers/tokenization_utils_base.py", line 2256, in _from_pretrained
> >     tokenizer = cls(*init_inputs, **init_kwargs)
> >   File "/home/dell/anaconda3/envs/vllm/lib/python3.8/site-packages/transformers/models/gpt2/tokenization_gpt2_fast.py", line 136, in __init__
> >     super().__init__(
> >   File "/home/dell/anaconda3/envs/vllm/lib/python3.8/site-packages/transformers/tokenization_utils_fast.py", line 111, in __init__
> >     fast_tokenizer = TokenizerFast.from_file(fast_tokenizer_file)
> > Exception: EOF while parsing a value at line 156568 column 3
> > ```
> 
> Both the `tokenizer.json` and the 7th model file are broken on huggingface, you need to download then from model.baai.ac.cn

we just updated the broken tokenizer files.

**Comment by gesanqiu** [comment]
> All right, I managed to solve the model loading issue. This is because in `aquila2` model, there is a config `num_key_value_heads`, just like `llama2`. Apply this patch as well as this pr, then `aquila2` model can run:
> 
> ```
> diff --git a/vllm/model_executor/models/aquila.py b/vllm/model_executor/models/aquila.py
> index 33280d9..8786287 100644
> --- a/vllm/model_executor/models/aquila.py
> +++ b/vllm/model_executor/models/aquila.py
> @@ -147,6 +147,7 @@ class AquilaAttention(nn.Module):
>              rotary_dim=self.head_dim,
>              base=self.rope_theta,
>              max_position=self.max_position_embeddings,
> +            num_kv_heads=self.num_kv_heads
>          )
>  
>      def forward(
> @@ -177,7 +178,7 @@ class AquilaDecoderLayer(nn.Module):
>          self.self_attn = AquilaAttention(
>              hidden_size=self.hidden_size,
>              num_heads=config.num_attention_heads,
> -            num_kv_heads=config.num_attention_heads,
> +            num_kv_heads=config.num_key_value_heads,
>              rope_theta=rope_theta,
>              max_position_embeddings=max_position_embeddings,
>          )
> @@ -308,7 +309,7 @@ class AquilaForCausalLM(nn.Module):
>          q_proj_shard_size = (self.config.hidden_size // tp_size)
>          kv_proj_shard_size = (self.config.hidden_size //
>                                self.config.num_attention_heads *
> -                              self.config.num_attention_heads // tp_size)
> +                              self.config.num_key_value_heads // tp_size)
>          attention_weight_specs = [
>              # (weight_name, shard_size, offset)
>              ("q_proj", q_proj_shard_size, 0),
> diff --git a/vllm/transformers_utils/configs/aquila.py b/vllm/transformers_utils/configs/aquila.py
> index 944e8f0..738a50f 100644
> --- a/vllm/transformers_utils/configs/aquila.py
> +++ b/vllm/transformers_utils/configs/aquila.py
> @@ -33,6 +33,7 @@ class AquilaConfig(PretrainedConfig):
>          intermediate_size=11008,
>          num_hidden_layers=32,
>          num_attention_heads=32,
> +        num_key_value_heads=None,
>          hidden_act="silu",
>          max_position_embeddings=2048,
>          initializer_range=0.006,
> @@ -49,6 +50,12 @@ class AquilaConfig(PretrainedConfig):
>          self.hidden_size = hidden_size
>          self.intermediate_size = intermediate_size
>          self.num_hidden_layers = num_hidden_layers
> +        
> +        # for backward compatibility
> +        if num_key_value_heads is None:
> +            num_key_value_heads = num_attention_heads
> +
> +        self.num_key_value_heads = num_key_value_heads
>          self.num_attention_heads = num_attention_heads
>          self.hidden_act = hidden_act
>          self.initializer_range = initializer_range
> ```

Is this patch compatible to the `Aquila1` model?

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thanks for the fix!

**Comment by zhuohan123** [comment]
@lu-wang-dl can you confirm when all the comments are addressed and the PR is ready to be merged?

**Comment by lonngxiang** [comment]
After merging the main branch, will the use test wait for vllm to release the new version? If you directly modify the place a little trouble

**Comment by exceedzhang** [comment]
I ran vLLM using AquilChat2-7B-16K model and found that the output was incorrect, as shown in the following figure：
image

**Comment by lonngxiang** [comment]
how about this problem ： https://github.com/vllm-project/vllm/issues/1349

**Comment by exceedzhang** [comment]
“Hello! What is your name?”
image

image

**Comment by exceedzhang** [comment]
image

**Comment by lonngxiang** [comment]
> “Hello! What is your name?” image
> 
> image

Thanks for sharing your screenshot of the proces

**Comment by exceedzhang** [comment]
> > I ran vLLM using AquilChat2-7B-16K model and found that the output was incorrect, as shown in the following figure： image
> 
> 16k model needs rope, and currently aquila model class from vllm haven't support this. Perhaps you can try port rope support from llama model class.

AquilaChat2-7B Work Well! Great job!
image

**Comment by lonngxiang** [comment]
and 

> > > I ran vLLM using AquilChat2-7B-16K model and found that the output was incorrect, as shown in the following figure： image
> > 
> > 
> > 16k model needs rope, and currently aquila model class from vllm haven't support this. Perhaps you can try port rope support from llama model class.
> 
> AquilaChat2-7B Work Well! Great job! image

Can you share your code to learn?

**Pull request merged**
